### PR TITLE
internal/debug: use pprof goroutine writer for debug_stacks

### PR DESCRIPTION
**Pull request by ryanschneider**
When debugging an issue today, I found that `debug.Stacks()` is limited to 1M of goroutines, and sure enough because of a bunch of wayward goroutines related to the `Feed.Send` calls discussed in #16673 the goroutine I was really interested wasn't included in the dump.

So, at first I was going to submit a simple fix like this, to dynamically resize the dump buffer (stolen from the go source for debug.Stack):

```
		buf := make([]byte, 1024*1204)
		for {
			n := runtime.Stack(buf, true)
			if n < len(buf) {
				return string(buf[:n])
			}
			buf = make([]byte, 2*len(buf))
		}
```

However, that felt dirty (and potentially very slow), so I started to look around if there was anything that deduped similar goroutines before dumping, and came across the older dump format in `pprof`, which does goroutine de-duping:

```
goroutine profile: total 148
16 @ 0x442edc 0x442fce 0x41a0b4 0x419d5b 0x986c85 0x984523 0x9842c2 0x983fea 0x98b30b 0x4728a1
#	0x986c84	github.com/ethereum/go-ethereum/p2p/discover.(*udp).ping+0x364			/ext-go/1/src/github.com/ethereum/go-ethereum/p2p/discover/udp.go:289
#	0x984522	github.com/ethereum/go-ethereum/p2p/discover.(*Table).ping+0xd2			/ext-go/1/src/github.com/ethereum/go-ethereum/p2p/discover/table.go:663
#	0x9842c1	github.com/ethereum/go-ethereum/p2p/discover.(*Table).pingpong+0xb1		/ext-go/1/src/github.com/ethereum/go-ethereum/p2p/discover/table.go:644
#	0x983fe9	github.com/ethereum/go-ethereum/p2p/discover.(*Table).bond+0x659		/ext-go/1/src/github.com/ethereum/go-ethereum/p2p/discover/table.go:616
#	0x98b30a	github.com/ethereum/go-ethereum/p2p/discover.(*Table).bondall.func1+0xda	/ext-go/1/src/github.com/ethereum/go-ethereum/p2p/discover/table.go:562
```

The first `16` means that 16 goroutines were found with a similar stack trace.

Here's a [full dump](https://gist.githubusercontent.com/ryanschneider/9c4b16c6779f6cd168230d6ee2913b59/raw/275e85a91f0112d774acf9be4a74e827bced95f0/debug.stacks.log) from a recently started node as well.

Personally, I find the new format very readable, but it is different than the old format, but I think that de-duping is really worthwhile.

### CODE REVIEW

**Comment by ryanschneider** [comment]
Hmm, but it doesn't include the goroutine state info (`runnable`, etc), which is also very valuable for debugging...

**Comment by ryanschneider** [comment]
However, changing the `1` to a `2` in .`WriteTo(..., 2)` outputs the old format and ensures that the buffer is big enough internally in the pprof code:

```golang
// /usr/local/go/src/runtime/pprof/pprof.go:602
func writeGoroutineStacks(w io.Writer) error {
	// We don't know how big the buffer needs to be to collect
	// all the goroutines. Start with 1 MB and try a few times, doubling each time.
	// Give up and use a truncated trace if 64 MB is not enough.
	buf := make([]byte, 1<<20)
	for i := 0; ; i++ {
		n := runtime.Stack(buf, true)
		if n < len(buf) {
			buf = buf[:n]
			break
		}
		if len(buf) >= 64<<20 {
			// Filled 64 MB - stop there.
			break
		}
		buf = make([]byte, 2*len(buf))
	}
	_, err := w.Write(buf)
	return err
}
```

So seems like the trade-off is:

- de-duplication
- goroutine state (`runnable`, `chan receive`, etc.)

Ideally we could find a format that provides both, though..

**Comment by ryanschneider** [comment]
If we can't, would you all be opposed to adding a param to `debug_stacks` RPC for `format`?  The default could be the current format.

**Comment by karalabe** [comment]
Hmm, the status of the goroutines is pretty much essential to debug locks, but the condensed format is light years cleaner. Couldn't we somehow hack/fix pprof to support dedup + status? Would be invaluable imho.

**Comment by fjl** [comment]
You can also get the pprof format from the pprof http endpoint (if you start with `--pprof`).

**Comment by ryanschneider** [comment]
Hmm, so the function for getting a goroutine's status header is here:

https://github.com/golang/go/blob/master/src/runtime/traceback.go#L867

But I haven't found anything that actually exposes that beyond `tracebackothers` in the same file, which is only called by `runtime.Stack(_, true)`.

So, I think the only recourse would be to do our own de-duping, similar to this project: https://github.com/linuxerwang/goroutine-inspect/blob/a1fe3461dd7c7b9b73541de733ed26df65c4eb73/goroutine.go#L232

However, IMO, if we went down this route it should probably be it's own module and vendored into geth since it wouldn't be geth-specific and could definitely have value in other apps.

For now, I think I'll just change my PR to use `2` so we are sure to get the full stack dump (and doing so using the `pprof` module so if that gets optimized in later go versions we get it for free).

**Comment by ryanschneider** [comment]
Ok, I've forced pushed an update to the branch to use mode 2 to keep the output format the same (but using pprof so we get the full output up to 64MB instead of the current 1MB cutoff).  Time permitting I will continue to look into de-duping the goroutines while keeping their state as a separate PR/module.

**Comment by ryanschneider** [comment]
@karalabe may I request that this is added to the 1.8.11 milestone?  I again hit an issue today where the current stack output was truncated.  To recap the current version uses the same output as before, but it is no longer capped to 1MB, and it uses a routine in the `pprof` package to generate the output, which for Go 1.10 looks very similar to our old code (but will grow the buffer up to 64MB as needed), and of course might be further optimized in a later Go release.

Separately I'm planning another PR that enables the alternative output (de-duped but w/o goroutine status) as either another debug RPC or an argument on debug_stacks, but I doubt I'll get to that before 1.8.11.

**Comment by fjl** [comment]
Sorry, this didn't make it into 1.8.11 but should be good to go now. Thank you!

**Comment by fjl** [comment]
What I meant earlier: if you want to get the pprof format for debugging, just start with `geth --pprof`. You can then use the pprof HTTP API at /debug/pprof. It's also very handy for remote profiling.

File: internal/debug/api.go [hunk 1]
```diff
@@ -190,9 +191,9 @@ func (*HandlerT) WriteMemProfile(file string) error {
 
 // Stacks returns a printed representation of the stacks of all goroutines.
 func (*HandlerT) Stacks() string {
-	buf := make([]byte, 1024*1024)
-	buf = buf[:runtime.Stack(buf, true)]
-	return string(buf)
+	var b bytes.Buffer
+	pprof.Lookup("goroutine").WriteTo(w, 2)
```
Comment (karalabe) [inline internal/debug/api.go]: @fjl `w` -> `b`.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 53e5166eb04a3a58a752f60fb56e17bb0967c8c1
  * Commits must be prefixed with the package(s) they modify

 * Commit: 0c022a77e157849c5133d274f414aaec974e0a32
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Pull request merged**